Contact modified to be validable and accessible

LivioCavallo opened the pull request:
### Summary of Changes
Added missing < label > to < select > element, for accessibility.
Replaced align attribute on < img > element with css styling (vertical-align : middle).
Added missing < dt > child element inside < dl > element

### Testing Instructions
Joomla! 3.7.4
Create 2 Contacts and populate Image and Position fields (in default category).
Create a menuitem of type "list Contacts in a category" (default category) and in "Contact Display options" set "Show Contact List" ("Contact Position" and "Image" are set to "show" already).
Browse to that page and select one of the contacts, so a contacts selector is shown.

### Expected result
Validating HTML (https://validator.w3.org/) and accessibility check (https://achecker.ca/checker/index.php) should succeed.

### Actual result
Accessibility error: select element missing an associated label.
HTML Validation errors: The align attribute on the img element is obsolete. Use CSS instead.  /   Element dl is missing a required child element.

### Documentation Changes Required
none

### Notes
I think this is not a bug, but this PR is a little help toward html conformance and accessibility.

LivioCavallo committed:
Merge remote-tracking branch 'upstream/staging' into staging

LivioCavallo committed:
Merge remote-tracking branch 'upstream/staging' into staging

LivioCavallo committed:
content: added label and dt

added missing label for accessibility and dt for html validation

LivioCavallo committed:
Merge remote-tracking branch 'upstream/staging' into content.validable-accessible

LivioCavallo committed:
Merge remote-tracking branch 'upstream/staging' into content.validable-accessible

matrikular commented on components/com_contact/views/contact/tmpl/default.php:
```diff
@@ -84,12 +84,13 @@
 		<?php if ($this->contact->image && $tparams->get('show_image')) : ?>
 			<div class="thumbnail pull-right">
-				<?php echo JHtml::_('image', $this->contact->image, $this->contact->name, array('align' => 'middle', 'itemprop' => 'image')); ?>
+				<?php echo JHtml::_('image', $this->contact->image, $this->contact->name, array('style' => 'vertical-align:middle;', 'itemprop' => 'image')); ?>
```
Please add a whitespace here (vertical-align: middle;).

LivioCavallo replied:
Sure.

matrikular commented on components/com_contact/views/contact/tmpl/default.php:
```diff
@@ -84,12 +84,13 @@
 
 		<?php if ($this->contact->con_position && $tparams->get('show_position')) : ?>
 			<dl class="contact-position dl-horizontal">
+				<dt><?php echo JText::_('COM_CONTACT_POSITION').': '; ?></dt>
```
While adding the < dt > tag is a good* idea, any addition here would also have to be made in all core templates that ship with an override for this layout. Apart from that, either add a whitespace between the concatenation (dot), or place the colon behind the PHP statement (prefered), please. `<dt><?php echo JText::_('COM_CONTACT_POSITION'); ?>:</dt>`

_*Adding the < td > tag might have a negative influence on the way this section of the page is displayed / handled._

LivioCavallo replied:
I agree, this should be done in all templates with overrides. May be I'll do that.

matrikular commented on components/com_contact/views/contact/tmpl/default.php:
```diff
@@ -53,8 +53,8 @@
 	<?php if ($tparams->get('show_contact_list') && count($this->contacts) > 1) : ?>
 		<form action="#" method="get" name="selectForm" id="selectForm">
-			<?php echo JText::_('COM_CONTACT_SELECT_CONTACT'); ?>
-			<?php echo JHtml::_('select.genericlist', $this->contacts, 'id', 'class="inputbox" onchange="document.location.href = this.value"', 'link', 'name', $this->contact->link); ?>
```
Choosing "id" might unarguably be the worst possible choice for an actual element ID, but I also see two to three things to improve your proposed changes with.
- First: The ID should not containt the prefix "idSel[...]". It feels like prefixing variables with their type (e.g. $arrSomeArray = [], var intAge = 38;).
Note: Without explicitly adding the ID as a parameter to the genericlist method, the name and id attribute will be equal. As a result, the field name would be "idSelContact".
- Second: Abbreviations. I remember a time when writing IDs that way was pretty common in ASP.NET. I may be "wrong", but personally I don't see much benefit in this. Especially not, when you need to refactor field names and their corresponding IDs and / or (for-)attributes.
- Third: Apart from a few exceptions (forms), the majority of IDs in the CMS are written in lowercase. There is no style guide yet that states all IDs must be written in lowercase, but I would suggest to do the same here. 

Long story short: select_contact

LivioCavallo replied:
Ok. Very reasonable. Thanks.
My intention was to mimic some Joomla style but evidently I missed the point...

matrikular reviewed (changes requested):
Please have a look at the review comments.

LivioCavallo committed:
Merge branch 'content.validable-accessible' into upstream/staging

LivioCavallo committed:
Merge remote-tracking branch 'upstream/staging' into content.validable-accessible

LivioCavallo committed:
Integrate matrikular suggestions

id name, spaces

LivioCavallo committed:
beez3: contact view updated (validation & accessibility)

Updated beez3 template contact view override to integrate fixes for accessibility and validation, as done in joomla core view

LivioCavallo committed:
moved space

LivioCavallo commented:
The same fix applied to beez3 template contact view override

matrikular commented on components/com_contact/views/contact/tmpl/default.php:
```diff
@@ -53,8 +53,8 @@
 		<form action="#" method="get" name="selectForm" id="selectForm">
-			<?php echo JText::_('COM_CONTACT_SELECT_CONTACT'); ?>
-			<?php echo JHtml::_('select.genericlist', $this->contacts, 'id', 'class="inputbox" onchange="document.location.href = this.value"', 'link', 'name', $this->contact->link); ?>
+			<?php echo '<label for="select_contact">'.JText::_('COM_CONTACT_SELECT_CONTACT').'</label>'; ?>
```
Please split the HTML markup from the PHP part.
`<label for="select_contact"><?php echo JText::_('COM_CONTACT_SELECT_CONTACT'); ?></label>`

For J!4 onward, I suggest we use a JForm field to be rendered here instead.

zwiastunsw replied:
Is this vertical-align style needed? The photo is placed inside a div element. Applying style does not change anything.

LivioCavallo replied:
I think it's not needed. Can anybody confirm, please?

uglyeoin commented:
Can this be tagged with accessibility please?

ghost commented:
done

LivioCavallo committed:
Merge remote-tracking branch 'upstream/staging' into content.validable-accessible

LivioCavallo committed:
Split html from php

matrikular commented:
I have tested this item :white_check_mark: successfully on d26b936f407d39a35ed9673af21c77fc787bc0e8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17602">issues.joomla.org/tracker/joomla-cms/17602</a>.</sub>

LivioCavallo commented:
I have tested this item :white_check_mark: successfully on d26b936f407d39a35ed9673af21c77fc787bc0e8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17602">issues.joomla.org/tracker/joomla-cms/17602</a>.</sub>

ghost commented:
@LivioCavallo can't count your Test as you created this Pull request and its expected that Dev. test their PRs.

matrikular commented on templates/beez3/html/com_contact/contact/default.php:
```diff
@@ -43,8 +43,8 @@
 		<form action="#" method="get" name="selectForm" id="selectForm">
-			<?php echo JText::_('COM_CONTACT_SELECT_CONTACT'); ?>
-			<?php echo JHtml::_('select.genericlist',  $this->contacts, 'id', 'class="inputbox" onchange="document.location.href = this.value"', 'link', 'name', $this->contact->link);?>
+			<?php echo '<label for="select_contact">'.JText::_('COM_CONTACT_SELECT_CONTACT').'</label>'; ?>
```
The beez layout is still mixing HTML / PHP markup.

LivioCavallo committed:
Updated beez3 template contact view override

LivioCavallo commented:
Can I do anything to help drone to solve "volume not found" error?

brianteeman commented:
I have restarted drone for you

zwiastunsw commented:
I have tested this item :white_check_mark: successfully on b6d5594ed1e7921a97710199ad3cb573f37c4346<br /><br />But I suggest you remove this vertical-align style. It's not needed.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17602">issues.joomla.org/tracker/joomla-cms/17602</a>.</sub>

LivioCavallo committed:
removed vertical-align: middle;

matrikular commented:
I have tested this item :white_check_mark: successfully on 3038d0923f264a55789091d09b27ecdcd1cede5d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17602">issues.joomla.org/tracker/joomla-cms/17602</a>.</sub>

zwiastunsw commented:
I have tested this item :white_check_mark: successfully on 3038d0923f264a55789091d09b27ecdcd1cede5d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17602">issues.joomla.org/tracker/joomla-cms/17602</a>.</sub>

ghost commented:
RTC after two successful tests.
